app/controllers/application_controller.rb

```diff
@@ -20,28 +19,52 @@ def jsonapi_render(collection, filters)
   end
 
   def set_user
-    @current_user = if auth_token.present?
-        User.from_jwt(Authenticator.from_token(auth_token))
-      else
-        nil
-      end
+    return nil unless auth_token.present?
+
+    @current_user = User.where(
+                      id: client.authenticated_user_id,
+                      login: client.authenticated_user_login
+                    ).first_or_create do |user|
+                      user.display_name = client.authenticated_user_display_name
+
+                      # Explicitly set user admin accessor if encoded in JWT
+                      user.admin = client.authenticated_admin?
+                    end
+
+    if needs_roles_refresh?
+      set_roles
+    end
   end
 
   def set_roles
     return unless current_user
-    current_user.roles = panoptes_client.roles current_user.id
+
+    current_user.update(roles: panoptes.roles(current_user.id), roles_refreshed_at: Time.now)
   end
 
-  def panoptes_client
-    @panoptes_client ||= PanoptesClient.new auth_token
+  def panoptes
+    @panoptes_api ||= PanoptesApi.new auth_token
+  end
+
+  def client
+    panoptes.client
   end
 
   def auth_token
     return @auth_token if @auth_token
+
     authorization = request.headers['Authorization']
     @auth_token = authorization.sub(/^Bearer /, '') if authorization.present?
   end
 
+  def needs_roles_refresh?
+    current_user.roles.nil? || current_user.roles_refreshed_at < token_created_at
+  end
+
+  def token_created_at
+   client.token_expiry - 2.hours
```

Comment:
would it be better to allow and env var with a default here ?
```suggestion
   client.token_expiry - ENV.fetch(TOKEN_VALIDITY_TIME, 2.hours)
```

Reply:
I would feel better about this if the time was also being defined somewhere in Panoptes. So long as you're cool with this change being included but relying on the default (not setting the env var), this is good.

Reply:
this is defined in the dookeeper gem defaults, we could explicitly set it i the API config if you'd like. feel free to PR it.